type Array<T> = { T }

export type LuauExecutionSessionTask_State = "STATE_UNSPECIFIED" | "QUEUED" | "PROCESSING" | "CANCELLED" | "COMPLETE" | "FAILED"

--- Present when the task execution fails. Contains details about the error that
--- caused the failure.
export type LuauExecutionSessionTask_Error = {
	--- An error code indicating the category of the error.
	code: "ERROR_CODE_UNSPECIFIED" | "SCRIPT_ERROR" | "DEADLINE_EXCEEDED" | "OUTPUT_SIZE_LIMIT_EXCEEDED" | "INTERNAL_ERROR",
	--- An error message containing more details about the error.
	message: string,
}

--- Present when the task execution succeeds. Contains the output of the
--- execution.
export type LuauExecutionSessionTask_Output = {
	--- Return values from the script that was run. Return values that are not
	--- JSON serializable (such as Data Model Instances) will be returned as
	--- nulls.
	results: Array<any>,
}

--[[
    A `LuauExecutionSessionTask` ("task" for short) executes a given Luau script
    in the context of a specific version of a place.

    The script may access and update the data model of the place, including
    invoking any module scripts. However, data model changes are local to the
    task and cannot be persisted.

    The script can also invoke engine APIs that read and/or modify data stored
    in the cloud, such as those for DataStores. Exercise caution when using such
    APIs.
]]
export type LuauExecutionSessionTask_Request = {
	--- The script to be run as part of this task.
	script: string,
}

export type LuauExecutionSessionTask = LuauExecutionSessionTask_Request & {
	--- The resource path of the luau execution session task.
	path: string,
	--- Time when this task was created. This string is formatted as a
	--- Timestamp.
	createTime: string,
	--- Time when this task's state last changed. This string is formatted as a
	--- Timestamp.
	updateTime: string,
	--- The user that created the API key that was used to create this task.
	user: string,
	--- The task's state. See the State enum for information about each possible
	--- value.
	state: LuauExecutionSessionTask_State,

	error: LuauExecutionSessionTask_Error?,
	output: LuauExecutionSessionTask_Output?,
}

--[[
    Represents a chunk of log messages generated by a `LuauExecutionSessionTask`.

    The amount of logs that is retained is limited. See the documentation of the
    `LuauExecutionSessionTask` resource for more details.

    Logs have the same retention time as the parent task.
]]
export type LuauExecutionSessionTaskLog = {
	luauExecutionSessionTaskLogs: Array<{
		--- The resource path of the luau execution session task log.
		path: string,
		--- A list of log messages generated by the task execution. Each call to the
		--- Luau `print()` function from the task's script's code results in a
		--- single list item here, even if the message itself contains newlines.
		messages: Array<string>,
	}>,
}

return nil
